build-support/native-image/README.md

```diff
@@ -0,0 +1,87 @@
+
+Note that all json resource files can be inspected and transformed with the `jq` command-line tool [^11]. This is what is done in the script in [^1].
+
+# Automatically generating a zinc native-image for your repo
+
+**NOTE:** This will allow creating a zinc native-image of code containing macros, but the image currently has to be manually regenerated whenever adding or modifying macros!
+
+The script in [^1] will run on OSX or Linux (the `ubuntu:latest` container on docker hub is known to work). The script can be run to test out native-image zinc compiles as follows:
+
+``` bash
+$ cd /your/pants/codebase
+$ NATIVE_IMAGE_EXTRA_ARGS='-H:IncludeResourceBundles=org.scalactic.ScalacticBundle' /path/to/your/pants/checkout/build-support/native-image/generate-native-image-for-pants-targets.bash ::
+```
+
+After a long bootstrap process, the arguments are forwarded to a pants invocation which runs with reflection tracing. The `NATIVE_IMAGE_EXTRA_ARGS` environment variable can be used to add any necessary arguments to the `native-image` invocation (the scalactic resource bundle is necessary for any repo using scalatest). The above will build an image suitable for all targets in the repo (`::`) for the current platform. The script will generate a different output file depending upon whether it is run on OSX or Linux.
+
+On OSX:
+
+``` bash
+$ mkdir -pv ~/.cache/pants/bin/zinc-pants-native-image/mac/10.13/0.0.15
+$ cp -v zinc-pants-native-Darwin ~/.cache/pants/bin/zinc-pants-native-image/mac/10.13/0.0.15/zinc-pants-native-image
+```
+
+On Linux:
+
+``` bash
+$ mkdir -pv ~/.cache/pants/bin/zinc-pants-native-image/linux/x86_64/0.0.15
+$ cp -v zinc-pants-native-Linux ~/.cache/pants/bin/zinc-pants-native-image/linux/x86_64/0.0.15/zinc-pants-native-image
+```
+
+The native-image can then be tested out with:
+
+``` bash
+$ PANTS_COMPILE_ZINC_JVM_OPTIONS='[]' ./pants --zinc-native-image --zinc-version=0.0.15 compile.zinc --execution-strategy=hermetic --no-incremental --cache-ignore test ::
+```
+
+*Note:* if the native-image build fails, and you see the following in the output:
+``` bash
+Caused by: java.lang.VerifyError: class scala.tools.nsc.Global overrides final method isDeveloper.()Z
+```
+
+Please re-run the script at most two more times. This can occur nondeterministically for some reason right now.
+
+## Updating the zinc native-image
+
+This is a developing story. Currently, the script will idempotently create or update a directory in the pwd named `generated-reflect-config/` with the results of the reflection tracing. This directory will contain 5 files -- 4 json config files, and one `BUILD` file. This directory can be checked in and updated over time -- subsequent runs of the script will never remove information from previous runs.
```

Comment:
So, the script currently has it fixed to `0.0.15`, but we should be able to cover any changes to underlying zinc code by updating the config we embed in our `org.pantsbuild:zinc-compiler_2.12` artifact (since #7506). What I was actually trying to say here (and what I will edit this to) was that the script doesn't have to be run over the whole repo each time any macro is added -- rather, it can be run only over all invalidated targets (for example). In any case, it's always possible to just blow away the config dir and have it regenerated -- you would just need to re-run the script over the whole repo again (although this need not occur all at once).